libcxx: std::ostream::sentry should be exported

jeremyd2019 opened the pull request:
Not for review.  Testing another idea for #140145

kikairoya committed:
[Cygwin][MinGW] Internal class in explicitly-instantiation-declarated template should be instantiated

In-code comment says "explicit instantiation decl of the outer class
doesn't affect the inner class" but this behavior seems MSVC
specific, MinGW-GCC and Cygwin-GCC does not.
Clang should honor gcc's behavior.

This change fixes std::string compatibilty and resolves strange link
error (statically linked), strange crash (dynamically linked) using
libstdc++ on Cygwin.

mstorsjo commented:
FYI, note that even if you include clang changes here, those won't be used by the libcxx CI build; the CI uses a prebuilt build of llvm-mingw - see `.github/workflows/libcxx-build-and-test.yaml`.

jeremyd2019 commented:
I know, I just started with #140145 for my testing, this pr is really meant for testing that no other platform is broken.  https://github.com/jeremyd2019/llvm-mingw/actions/runs/15075388808 is running the full test of [af82a52](https://github.com/llvm/llvm-project/pull/140169/commits/af82a52986d549e32a1384630d4a26b8ecfa7687)

jeremyd2019 committed:
CI: run windows jobs without waiting for stage2

use test llvm-mingw toolchain

kikairoya committed:
[libc++] Introduce a new attribute keyword for Clang improves compatibility with Mingw-GCC

In MinGW environment, Clang handles dllexport attribute of internal
class that defined in class template in different way from GCC.
This incompatibility should be fixed but breaks ABI of libc++, so
introduce a new keyword to keep ABI in MinGW environment with
old and patched Clang and to stay ABI compatible on other platforms.

This attribute is attached only for basic_ostream::sentry::sentry,
basic_ostream::sentry::~sentry and basic_istream::sentry::sentry.
Other entities won't be affected by patching Clang so doesn't need
to be annotate.

At a time to introduce a new class as a non-template inner type of
a class template, all non-inline member functions of that class
also needs to be attached _LIBCPP_HIDE_FROM_ABI_MINGW_OR_AFTER_V1
or _LIBCPP_HIDE_FROM_ABI, not _LIBCPP_HIDE_FROM_ABI_AFTER_V1.
Otherwise, that member functions contained in DLL will be
inaccessible on MinGW environment.

For time to increase ABI version, _LIBCPP_HIDE_FROM_ABI_MINGW_OR_AFTER_V1
can be simply replaced to _LIBCPP_HIDE_FROM_ABI_AFTER_V1.

jeremyd2019 committed:
move inline keyword

jeremyd2019 committed:
[CI] test unstable abi in stage1

Since it seems so hard to get to stages 2 and 3
